tests/distributed/test_custom_all_reduce.py

```diff
@@ -90,24 +89,23 @@ def eager_allreduce(tp_size, pp_size, rank, distributed_init_port):
     num_communication = rank // tp_size + 1
     sz = 1024
     fa = get_tp_ca_communicator()
-    inp = torch.ones(sz, dtype=torch.float32, device=device)
-    out = inp
-    for _ in range(num_communication):
-        out = fa.all_reduce_unreg(out)
-    assert torch.allclose(out, inp * (tp_size**num_communication))
-
-    inp = torch.ones(sz * 4, dtype=torch.bfloat16, device=device)
+    inp = torch.ones(sz, dtype=dtype, device=device)
     out = inp
     for _ in range(num_communication):
         out = fa.all_reduce_unreg(out)
     assert torch.allclose(out, inp * (tp_size**num_communication))
 
 
+@pytest.mark.skipif(torch.cuda.device_count() < 2,
+                    reason="Need at least 2 GPUs to run the test.")
 @pytest.mark.parametrize("tp_size", [2])
 @pytest.mark.parametrize("pipeline_parallel_size", [1, 2])
 @pytest.mark.parametrize("test_target", [eager_allreduce, graph_allreduce])
-def test_custom_allreduce(tp_size, pipeline_parallel_size, test_target):
+@pytest.mark.parametrize("dtype",
+                         [torch.float32, torch.float16, torch.bfloat16])
+def test_custom_allreduce(tp_size, pipeline_parallel_size, test_target, dtype):
     world_size = tp_size * pipeline_parallel_size
     if world_size > torch.cuda.device_count():
         pytest.skip("Not enough GPUs to run the test.")
-    multi_process_tensor_parallel(tp_size, pipeline_parallel_size, test_target)
+    multi_process_tensor_parallel(tp_size, pipeline_parallel_size, test_target,
+                                  dtype)
```

Comment:
it would be better if we can decouple the arg of `multi_process_tensor_parallel` and the arg of `test_target`, e.g. passing a wrapper function to ray.

Reply:
Thanks for your advice. I tried, but the implemented code seemed to be complicated and a little ugly because tp_size, pp_size, rank, and distributed_init_port are used by both multi_process_tensor_parallel and test_target. In the end, I think this PR is clearer and easier to read. If you have a better reference, I'll try again. Thank you a lot :)